r2-core/src/main/java/com/linkedin/r2/transport/http/client/CallbackWrappingConstantQpsRateLimiterFactory.java

```diff
@@ -0,0 +1,34 @@
+package com.linkedin.r2.transport.http.client;
+
+import com.linkedin.r2.transport.http.client.ratelimiter.CallbackWrapper;
+import com.linkedin.util.clock.Clock;
+import java.time.temporal.ChronoUnit;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ScheduledExecutorService;
+
+
+public class CallbackWrappingConstantQpsRateLimiterFactory implements ConstantQpsRateLimiterFactory {
```

Comment:
What is this RateLimiterFactory used for? Only for test purpose? Then can we make this an inner class of the test file

Reply:
Glad you asked! Before this change, we'd pass in a ConstantQpsRateLimiter that we could extend with whatever additional functionality we desired through an inheritance programming model. The move to supply a ConstantQpsRateLimiterFactory changes this to a composition model, where we're now instead passing the functionality we need in the form of a wrapper.

Name dropping class names from LI internal libraries: We'll no longer be able to use ICAwareCallbackCircularBuffer in the package that pulls in rest.li, and will instead be passing a ICAwareCallbackWrapper.

Reply:
RateLimiterFactory is the essential change in this PR. By supplying a RateLimiterFactory, it grants rest.li the ability to create its own RateLimiters, and it's able to create a separate RateLimiter for every downstream dark cluster. Without this change, rest.li only has the one rate limiter, so it has all downstream dark clusters sharing that one rate limiter. This gets pretty gnarly when each ConstantQpsDarkClusterStrategy independently calls setRate (for example) on that rate limiter.

Reply:
ConstantQpsRateLimiterFactory has been removed in the latest iteration of the PR. Instead, the API takes a Supplier of the same rate limiter to accomplish the same goal while requiring no new classes.